lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs

```diff
@@ -366,13 +396,18 @@ impl CompilerData {
         self.with_next_job("shader compilation", (|t| &mut t.jobs.compile), |this, job| {
             let profiler = job.profiler;
             profiler.resume();
-            let vertex = this.compile_shader(Vertex, job.input.vertex)?;
-            let fragment = this.compile_shader(Fragment, job.input.fragment)?;
+            let vertex = this.cache.get_or_insert_vertex_shader(job.cache_key, || {
+                CompilerData::compile_shader(&this.context, Vertex, job.input.vertex)
+            })?;
+            let fragment = this.cache.get_or_insert_fragment_shader(job.cache_key, || {
+                CompilerData::compile_shader(&this.context, Fragment, job.input.fragment)
+            })?;
```

Comment:
In both functions you are passing `Vertex` and `Fragment` values. Based on them, the function `get_or_insert_*_shader` could decide whether it is about vex or frag shader, right? Or is it super-complex to be done here?

Reply:
This is possible with some extra trait impls on the cache, but it absolutely isn't worth it for two cases. I'm still going to simplify those methods by taking their common part into a `HashMapOps` extension trait, as it is a very generic operation.